Code/Framework/AzCore/AzCore/std/containers/span.h

```diff
@@ -132,6 +128,24 @@ namespace AZStd
         pointer m_begin;
         pointer m_end;
     };
+
+    template<class Container>
+    span(const Container&) -> span<const typename Container::value_type>;
+
+    template<class Container>
+    span(Container&) -> span<typename Container::value_type>;
+
+    template<class Element>
+    span(typename span<Element>::const_pointer, typename span<Element>::size_type) -> span<const typename span<Element>::value_type>;
+
+    template<class Element>
+    span(typename span<Element>::pointer, typename span<Element>::size_type) -> span<typename span<Element>::value_type>;
+
+    template<class Element>
+    span(typename span<Element>::const_pointer, typename span<Element>::const_pointer) -> span<const typename span<Element>::value_type>;
+
+    template<class Element>
+    span(typename span<Element>::pointer, typename span<Element>::pointer) -> span<typename span<Element>::value_type>;
```

Comment:
The `typename span<Element>` can't be used in a deduction guide, as it requires types dependent on the template itself.


The parameters and return cannot to not depend on the `span` template
So what would need to be done is something like the following
```
template<class It, class EndOrSize>
span(It, EndOrSize) -> span<std::remove_reference_t<std::iter_reference_t<It>>>;
```

Reply:
I suggest you wait on adding a deduction guide, because this will not compile using clang.

Furthermore we don't have all of the type traits needed to implement span properly yet.

Reply:
@lumberyard-employee-dm solved this offline - the deduction rules were unnecessary, the types themselves just needed to be changed. Also, this led to the discovery that vector.h had redundant validate_iterator definitions that could lead to compile errors.  Changes are now posted in this PR.